✨ Hollow Machines in CAPD

fabriziopandini opened the pull request:
**What this PR does / why we need it**
This PR proposes an implementation of HollowMachines in Cluster API, based on kubemark/hollow kubelet.

The idea was originally proposed by @benmoss, but I took a different path in implementation so the delta diff is minimal and hopefully much easer to maintain…

**Which issue(s) this PR fixes:**
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/4565

enxebre commented on test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go:
```diff
@@ -0,0 +1,113 @@
+)
+
+const (
+	// HollowMachineFinalizer allows Reconcileto clean up resources associated with HollowMachine before
```
typo `Reconcileto`

enxebre commented on test/infrastructure/docker/controllers/hollowmachine_controller.go:
```diff
@@ -0,0 +1,221 @@
+	// Fetch the Cluster.
+	cluster, err := util.GetClusterFromMetadata(ctx, r.Client, machine.ObjectMeta)
+	if err != nil {
+		log.Info("HollowMachine owner HollowKubelet is missing cluster label or cluster does not exist")
```
should this log as error?

fabriziopandini replied:
we are alrading returning the error , which cause log.Error + exponential backoff

enxebre commented on test/infrastructure/docker/controllers/hollowmachine_controller.go:
```diff
@@ -0,0 +1,221 @@
+	}
+	// Always attempt to Patch the HollowMachine object and status after each reconciliation.
+	defer func() {
+		// TODO: investigating patching errors on delete
```
Personally I find the defer/patch/rerr pattern more difficult to follow than being explicit.
Why `if rerr == nil {`? `rerr` seems to be always nil.

fabriziopandini replied:
defer is the standard approach in CAPI controllers.
`rerr` value depends on the return statements down in the same function

fabriziopandini commented:
/hold
while discussing if/how to consolidate with https://github.com/kubernetes-sigs/cluster-api-provider-kubemark

elmiko commented:
i think this is cool, but i would like to maximize our efforts around kubemark. whether in capd or its own provider, i think it will be valuable.

for me, the main difference between this implementation and the standalone kubemark provider is that the latter presents an option for deploying onto non-capd clusters. otherwise it seems like _functionally_ they would be the same.

elmiko commented on test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go:
```diff
@@ -0,0 +1,113 @@
+	// Compute Resources required by kubemark.
+	// if not specified cpu and memory requests, defaults of 40m and 10240Ki will be appied.
+	// +optional
+	Resources corev1.ResourceRequirements `json:"resources,omitempty"`
```
this is the resource requirements to run kubemark?

fabriziopandini replied:
Yes, those are resources for the container running the hollowkubelet

sbueringer commented:
I took a rough look, looks very promising!

elmiko commented:
this generally makes sense to me, and i agree with @sbueringer very promising!

i am currently trying to test this locally but hitting networking issues, i don't think it's related to the PR but i will report back if i cannot solve these issues.

fabriziopandini commented:
/hold cancel
we agreed on keep developing this in CAPD to start exploring this space with SIG scalability

elmiko commented:
been doing some more hacking on this, things are going well but i hit a small snag today when i updated my cluster-api code. i needed to rebase this patch, which required the following minor update:
```
diff --git a/test/infrastructure/docker/main.go b/test/infrastructure/docker/main.go
index afbb9c496..2a3b39043 100644
--- a/test/infrastructure/docker/main.go
+++ b/test/infrastructure/docker/main.go
@@ -33,7 +33,6 @@ import (
        "k8s.io/klog/v2"
        "k8s.io/klog/v2/klogr"
        clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
-       "sigs.k8s.io/cluster-api/controllers/remote"
        expv1 "sigs.k8s.io/cluster-api/exp/api/v1alpha4"
        "sigs.k8s.io/cluster-api/feature"
        infrav1old "sigs.k8s.io/cluster-api/test/infrastructure/docker/api/v1alpha3"
@@ -152,8 +151,8 @@ func setupReconcilers(ctx context.Context, mgr ctrl.Manager) {
        // Set up a ClusterCacheTracker to provide to controllers
        // requiring a connection to a remote cluster
        tracker, err := remote.NewClusterCacheTracker(
-               ctrl.Log.WithName("remote").WithName("ClusterCacheTracker"),
                mgr,
+               remote.ClusterCacheTrackerOptions{Log: ctrl.Log.WithName("remote").WithName("ClusterCacheTracker")},
        )
        if err != nil {
                setupLog.Error(err, "unable to create cluster cache tracker")
```

i'm hopeful to finish up my testing in the next few days, but i'm very optimistic about this work =)

fabriziopandini commented:
Rebased!

elmiko commented on test/infrastructure/docker/api/v1alpha4/hollowmachine_types.go:
```diff
@@ -0,0 +1,159 @@
+
+	// ExternalCluster define the name of the cluster where Pods hosting kubemark should be executed.
+	// If empty, the machine's cluster will be used.
+	ExternalCluster *string `json:"externalCluster"`
```
does this need an `omitempty` added to allow for the defaulting behavior?

elmiko reviewed (commented):
just a question about one of the fields, also i think it would be really nice to have a readme with an example showing how to create a docker cluster that uses hollow nodes.

elmiko commented:
i've been testing with some cluster templates that @fabriziopandini shared with me, and i'm hitting a snag that i'm having trouble debugging.

when i launch the external cluster (where the pods for the hollow cluster will run), i see it go into provisioning and it never becomes provisioned. i looked into the capd controller and saw a bunch of these errors at the end:

```
E0708 21:24:18.559716       1 controller.go:302] controller-runtime/manager/controller/dockermachine "msg"="Reconciler error" "error"="failed to pre-load images into the DockerMachine: command \"docker save -o /tmp/image-tar3286373947/image-0.tar gcr.io/k8s-staging-cluster-api/kubemark:v1.20.2\" failed with error: exit status 1" "name"="external-cluster-1-control-plane-tk7hv" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" 
```

is this just a local configuration error?

do i need to pre-load some images into my local docker registry?

sbueringer commented:
@elmiko I would have guessed you need the `gcr.io/k8s-staging-cluster-api/kubemark:v1.20.2` image in your local Docker engine. I got a very similar error yesterday when I forgot to build the manager images before running the tests, seems to say more or less "this local image is missing".

elmiko commented:
ack, thanks @sbueringer 

i will try loading these into the local docker registry.

elmiko commented:
had to build my own kubemark image, and load it locally. i couldn't access the one at gcr.io. so far, so good though, making progress.

elmiko commented:
seem to be hitting a new error on the workload cluster, it gets stuck in provisioning and i see this in the capd-controller logs
```
E0709 18:27:13.812286       1 controller.go:302] controller-runtime/manager/controller/dockermachine "msg"="Reconciler error" "error"="failed to pre-load images into the DockerMachine: failed to load image: command \"docker exec --privileged -i test-1-control-plane-kxlr9 ctr --namespace=k8s.io images import -\" failed with error: exit status 1" "name"="test-1-control-plane-ddp97" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" 
```

fabriziopandini commented:
rebased
@elmiko I have also added examples of yaml for creating the external cluster and the test cluster with comments on the hollow specific bits. Hopefully this will help you in getting started in your environment

elmiko commented:
been test driving the latest changes, and when i try to spawn a cluster from the `hollow-machines-external-cluster.yaml` i see this error after doing `kubectl apply`:

```Error from server (InternalError): error when creating "capd-hollow-cluster/hollow-machines-external-cluster.yaml": Internal error occurre
d: failed calling webhook "default.clusterresourceset.addons.cluster.x-k8s.io": the server could not find the requested resource```

i'm still digging into it, but just curious if anyone has seen this before?

fabriziopandini commented:
> failed calling webhook "default.clusterresourceset.addons.cluster.x-k8s.io"

You need to turn on the CRS feature flag in your test env

elmiko commented:
i was able to get past the CRS feature error and have hit a new problem. it appears that when i create the test cluster (the full kubemark cluster that will attach to the external cluster), i am seeing errors in the capd-controller logs that look like an open file error:

**capd controller logs**
```
I0720 20:25:29.046302       1 machine.go:476] controller-runtime/manager/controller/dockermachine "msg"="Got logs from the machine container" "name"="hollow-cluster-control-plane-t4xlt" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" "output"=["INFO: ensuring we can execute mount/umount even with userns-remap","INFO: remounting /sys read-only","INFO: making mounts shared","INFO: detected cgroup v2","INFO: clearing and regenerating /etc/machine-id","Initializing machine ID from random ge
nerator.","INFO: faking /sys/class/dmi/id/product_name to be \"kind\"","INFO: faking /sys/class/dmi/id/product_uuid to be random","INFO: faking /sys/devices/virtual/dmi/id/product_uuid as well","INFO: setting iptables to detected mode: legacy","INFO: Detected IPv4 address: 172.18.0.8","INFO: Detected IPv6 address: ","systemd 246.6-1ubuntu1.1 running in system mode. (+PAM +AUDIT +SELINUX +IMA +APPARMOR +SMACK +SYSVINIT +UTMP +LIBCRYPTSETUP +GCRYPT +GNUTLS +ACL +XZ +LZ4 +ZSTD +SECCOMP +BLKID +ELFUTILS +KMOD +IDN2 -IDN +PCRE2 default-hierarchy=hybrid)","Detected virtualization docker.","Detected architecture x86-64.","","Welcome to \u001b[1mUbuntu 20.10\u001b[0m!","","Set hostname to <hollow-cluster-control-plane-6vgd7>.","\u001b[0;1;31mFailed to create control group inotify object: Too many open files\u001b[0m","\u001b
[0;1;31mFailed to allocate manager object: Too many open files\u001b[0m","[\u001b[0;1;31m!!!!!!\u001b[0m] Failed to allocate manager object.","\u001b[0;1;31mExiting PID 1...\u001b[0m"]
E0720 20:25:29.046638       1 controller.go:304] controller-runtime/manager/controller/dockermachine "msg"="Reconciler error" "error"="failed to exec DockerMachine bootstrap: failed to run cloud config: command \"docker exec --privileged hollow-cluster-control-plane-6vgd7 mkdir -p /etc/kubernetes/pki\" failed with error: exit status 1" "name"="hollow-cluster-control-plane-t4xlt" "namespace"="default" "reconciler group"="infrastructure.cluster.x-k8s.io" "reconciler kind"="DockerMachine" 
```

i thought maybe my system was setup with a low limit on these, but i double checked and it seems like it should be unlimited:
```
$ cat /proc/sys/fs/file-max 
9223372036854775807
$ ulimit
unlimited
```
and my docker service file seems to have unlimited as well:
```
LimitNOFILE=infinity
```

i'm not sure why it won't let that hostname file go through during bootstrap. i'm going to keep poking around to see if i can find more answers, but i figured this was worth raising to see if i am missing something.

vincepri commented:
@fabriziopandini Should we try to finish this PR before next release?

fabriziopandini commented:
@vincepri rebased
The PR is pretty stable and comments are about documentation/problems @elmiko local env. 
Personally I'm +1 to move this to the next stage and start trying to use it in some of our CI tests

elmiko commented:
+1 to what @fabriziopandini is saying, i am very excited to see this get merged. i am struggling to get it working locally, so i would like to make sure it easy for others to reproduce this workflow. i think that's my main concern currently.

fabriziopandini commented:
/close
As per office hours/slack discussions this will land in https://github.com/kubernetes-sigs/cluster-api-provider-kubemark
